lib/upload_to_s3.js

```diff
@@ -1,45 +1,81 @@
 import https from 'https';
 import url from 'url';
+import fs from 'fs';
+import temporary from 'temporary';
+import promiseRetry from 'promise-retry';
 
 export default async function uploadToS3 (task, sourceStream, artifactName, expiration, httpsHeaders) {
-  var queue = task.runtime.queue;
-
-  var artifact = await queue.createArtifact(
-    task.status.taskId,
-    task.runId,
-    artifactName,
-    {
-      // Why s3? It's currently cheaper to store data in s3 this could easily
-      // be used with azure simply by changing s3 -> azure.
-      storageType: 's3',
-      expires: new Date(Math.min(expiration, new Date(task.task.expires))),
-      contentType: httpsHeaders['content-type']
-    }
-  );
-
-  let parsedUrl = url.parse(artifact.putUrl);
-  let options = {
-    hostname: parsedUrl.hostname,
-    path: parsedUrl.path,
-    method: 'PUT',
-    headers: httpsHeaders
-  };
-
-  let req = https.request(options);
-  sourceStream.pipe(req);
-
-  let response = await new Promise((accept, reject) => {
-    req.on('response', accept);
-    req.on('error', reject);
-    req.setTimeout(5 * 60 * 1000, reject);
+  let queue = task.runtime.queue;
+
+  var artifact;
+  var parsedUrl;
+  var options;
+
+  let tmp = new temporary.File();
+  await new Promise((accept, reject) => {
+    let stream = fs.createWriteStream(tmp.path);
+    stream.on('error', reject);
+    stream.on('finish', accept);
+    sourceStream.pipe(stream);
   });
 
-  // Flush the data from the reponse so it's not held in memory
-  response.resume();
+  try {
+    await promiseRetry((retry, number) => {
+      if (number > 1) {
+        task.runtime.log('retrying artifact upload');
+      }
+
+      return new Promise(async (accept, reject) => {
+        new Promise((accept, reject) => {
+          if (artifact) {
+            accept();
+          } else {
+            queue.createArtifact(
```

Comment:
All in all, it makes no harm and testing gets a little bit easier if `createArtifact` is part of the retry process.